File: ChoirManagementApp/src/services/songService.ts
// src/services/songService.ts
import apiClient from './apiClient';
import { SongRequestDto, SongResponseDto, Page, PageableParams } from '../types/api';

export const SongApiService = {
  createSong: async (songData: SongRequestDto): Promise<SongResponseDto> => {
    const response = await apiClient.post<SongResponseDto>('/songs', songData);
    return response.data;
  },

  getAllSongs: async (title?: string, pageable?: PageableParams): Promise<Page<SongResponseDto>> => {
    const params: any = { ...pageable };
    if (title) {
      params.title = title;
    }
    const response = await apiClient.get<Page<SongResponseDto>>('/songs', { params });
    return response.data;
  },

  getSongById: async (id: number): Promise<SongResponseDto> => {
    const response = await apiClient.get<SongResponseDto>(`/songs/${id}`);
    return response.data;
  },

  updateSong: async (id: number, songData: SongRequestDto): Promise<SongResponseDto> => {
    const response = await apiClient.put<SongResponseDto>(`/songs/${id}`, songData);
    return response.data;
  },

  deleteSong: async (id: number): Promise<void> => {
    await apiClient.delete(`/songs/${id}`);
  },
};

File: my-spring-project/src/main/java/com/example/myspringproject/service/SongService.java
package com.example.myspringproject.service;

import com.example.myspringproject.dto.SongRequestDto;
import com.example.myspringproject.dto.SongResponseDto;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.Optional;

public interface SongService {

    SongResponseDto createSong(SongRequestDto songDto);

    Page<SongResponseDto> getAllSongs(String title, Pageable pageable);

    Optional<SongResponseDto> getSongById(Long id);

    Optional<SongResponseDto> updateSong(Long id, SongRequestDto songDto);

    void deleteSong(Long id);
}
